map2loop/thickness_calculator.py

```diff
@@ -340,3 +347,208 @@ def compute(
+
+        sampled_structures = geopandas.GeoDataFrame(
+            sampled_structures,
+            geometry=geopandas.points_from_xy(sampled_structures.X, sampled_structures.Y),
+            crs=basal_contacts.crs,
+        )
+        sampled_structures['unit_name'] = geopandas.sjoin(sampled_structures, geology)['UNITNAME']
+        sampled_basal_contacts = rebuild_sampled_basal_contacts(basal_contacts, sampled_contacts)
+
+        basal_contacts_bu = basal_contacts.copy()
+        basal_contacts_bu['geometry'] = basal_contacts_bu.buffer(0.1)
+
+        map_dx = geology.total_bounds[2] - geology.total_bounds[0]
+        map_dy = geology.total_bounds[3] - geology.total_bounds[1]
+
+        ths = []
+        lis = []
+
+        for s in range(0, len(sampled_structures)):
+            th = None
+            measurement = sampled_structures.iloc[s]
+            measurement_pt = shapely.Point(measurement.X, measurement.Y)
+            litho_in = measurement['unit_name']
+            strike = (measurement['DIPDIR'] - 90) % 360
+
+            bbox_poly = geology[geology['UNITNAME'] == litho_in][['minx', 'miny', 'maxx', 'maxy']]
+
+            GEO_SUB = geology[geology['UNITNAME'] == litho_in]['geometry'].values[0]
+            neighbour_list = list(
+                basal_contacts[GEO_SUB.intersects(basal_contacts.geometry)]['basal_unit']
+            )
+            B = calculate_endpoints(measurement_pt, strike, 10000, bbox_poly)
+            b = geopandas.GeoDataFrame({'geometry': [B]}).set_crs(basal_contacts.crs)
+
+            all_intersections = sampled_basal_contacts.overlay(
+                b, how='intersection', keep_geom_type=False
+            )
+            all_intersections = all_intersections[
+                all_intersections['geometry'].geom_type == 'Point'
+            ]
+
+            final_intersections = all_intersections[
+                all_intersections['basal_unit'].isin(neighbour_list)
+            ]
+
+            if 'MultiPoint' in final_intersections['geometry'].geom_type.values:
+                multi = final_intersections[
+                    final_intersections['geometry'].geom_type == 'MultiPoint'
+                ].index
+                for m in multi:
+                    nearest_ = shapely.ops.nearest_points(
+                        final_intersections.loc[m, :].geometry, measurement_pt
+                    )[0]
+                    final_intersections.at[m, 'geometry'] = nearest_
+                    final_intersections.at[m, 'geometry'] = nearest_
+
+            if len(final_intersections) < 2:
+                # print("Measurement n. ", s, "in litho ", litho_in, " has not enough intersections")
+                continue
+
+            if len(final_intersections['basal_unit'].unique()) == 1:
+                # print("Measurement n. ", s, "in litho ", litho_in, "does not cross two lithologies")
+                continue
+
+            int_pt1 = final_intersections.iloc[0].geometry
```

Comment:
Using pygeos you can get the coordinates of a geoseries, so the interesection points can be returned as a numpy array